src/main/kotlin/org/cascadebot/bot/cmd/HelpCommand.kt

```diff
@@ -0,0 +1,324 @@
+import org.cascadebot.bot.db.entities.CustomCommandEntity
+import org.cascadebot.bot.db.entities.GuildSlotEntity
+import org.cascadebot.bot.utils.QueryUtils.queryEntity
+import org.cascadebot.bot.utils.tryOrNull
+import java.util.UUID
+
+class HelpCommand : RootCommand("help", "Give you a list of all commands available to be run") {
+    override val commandData: CommandData = Commands.slash(name, description)
+        .addOptions(OptionData(OptionType.STRING, "command", "Command to get help on", false))
+
+    override fun onCommand(context: CommandContext, args: CommandArgs) {
+        if (args.getAmount() == 0) {
+            handleNoArgs(context)
+        } else {
+            handleCommandHelp(context, args.getArgAsString("command")!!)
+        }
+    }
+
+    private fun handleNoArgs(context: CommandContext) {
+        // TODO spilt into 3 pages, root commands, custom commands, and modules, for now their all going to be on one page
+        val globalBuilder = StringBuilder()
+        context.jda.retrieveCommands().queue { commands ->
+            for (command in commands) {
+                val cascadeCommand = Main.commandManager.getCommand(command.name)
+                val slashCommandData = cascadeCommand?.commandData as SlashCommandData
+                globalBuilder.append("* </${command.name}:${command.id}> ")
+                for (option in slashCommandData.options) {
+                    if (option.isRequired) {
+                        globalBuilder.append("<${option.name}> ")
+                    } else {
+                        globalBuilder.append("[${option.name}] ")
+                    }
+                }
+                globalBuilder.append('\n')
+            }
+            val guildBuilder = StringBuilder()
+            context.guild.retrieveCommands().queue { guildCommands ->
+                for (slot in Main.postgres.transaction {
+                    queryEntity(GuildSlotEntity::class.java) { root ->
+                        equal(root.get<Long>("guildId"), context.guild.idLong)
+                    }.list()
+                }) {
+                    for (guildCommand in guildCommands) {
+                        val customCommand = Main.postgres.transaction {
+                            return@transaction tryOrNull {
+                                queryEntity(CustomCommandEntity::class.java) { root ->
+                                    and(
+                                        equal(root.get<UUID>("slotId"), slot.slotId),
+                                        equal(root.get<String>("name"), guildCommand.name)
+                                    )
+                                }.singleResult
+                            }
+                        }
+                        if (customCommand == null) {
+                            // TODO remove from discord as it does not exist in db?
+                            continue
+                        }
```

Comment:
I think we don't do anything here in the help command - we could make that a process elsewhere

Reply:
I agree, but once we make that system this can be one of the places we call it from (probably best to do it async as to not delay the help command)